### PR TITLE
Pbi 741a

### PR DESCRIPTION
**Pull request by aeng7**
-Vast 3.0 podded ads functionality
-supports linear podded ads (tested with a sample tag) and theoretically supports podded ads that end with a nonlinear ad (untested due to lack of test tag)
-removed some wrapper ads functionality and unit test due to them functioning incorrectly before. Will have to be re-added once we tackle wrapper ads

### CODE REVIEW

**Comment by aeng7** [comment]
Please merge after Mjolnir PR 497 has been merged

**Comment by aeng7** [comment]
Tested mocked XML with non-linear ad at the end of podded ad. Worked fine.

**Comment by aeng7** [comment]
Do we want to add logic where if we find a non-linear overlay in the middle of a podded ad sequence, we push it to the end? The Vast 3.0 spec explicitly says that the non-linear overlay must be at the end of a podded ad sequence.

**Comment by aeng7** [comment]
Will add modified logic to handle ending ad pod before a non linear overlay if one exists. Also will change it so that ad count in pod will not count non linear ads. Will also add unit tests for these. Please do not merge yet.

**Comment by aeng7** [comment]
Ok to merge after review.

File: js/ad_manager_vast.js [hunk 1]
```diff
@@ -935,41 +993,113 @@ OO.Ads.manager(function(_, $) {
...
+     * @method Vast#handleAds
+     * @param {Array} ads An array of ad objects
+     * @param {object} adLoaded The ad loaded object and metadata
+     */
+    var handleAds = _.bind(function(ads, adLoaded) {
+      //find out how many non linear ads we have so as to not count them
+      //when determining ad pod length
+      var nonLinearAdCount = 0;
+      _.each(ads, _.bind(function(ad) {
+        if (!_.isEmpty(ad.nonLinear)) {
```
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: I think this will always evaluate to "not empty" because ad.nonLinear will always have "tracking" as a property like so:
```
ad.nonLinear = {
  tracking: {}
}
```
I think this happens because of LINE 789. double check me on that.

File: js/ad_manager_vast.js [hunk 2]
```diff
@@ -722,29 +769,28 @@ OO.Ads.manager(function(_, $) {
...
+        ad.linear.ClickTracking = wrapperAds.linear.ClickTracking.concat(ad.linear.ClickTracking || []);
+      }
+      if (wrapperAds.linear.tracking) {
+        if (!ad.linear.tracking) { ad.linear.tracking  = {}; }
+        _.each(wrapperAds.linear.tracking, function(value, key) {
+          ad.linear.tracking[key] = ad.linear.tracking[key] ? value.concat(ad.linear.tracking[key]) : value;
+        });
+      }
+      if (wrapperAds.nonLinear.tracking) {
+        if (!ad.nonLinear.tracking) { ad.nonLinear.tracking = {}; }
```
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: This is in reference to comment on LINE 1071.

File: js/ad_manager_vast.js [hunk 3]
```diff
@@ -935,41 +993,113 @@ OO.Ads.manager(function(_, $) {
...
+        podded : [],
+        standalone : []
+      };
+      //parse the ad objects from the XML
+      var ads = this.parseAds(vastXML);
+      //check to see if any ads are sequenced (are podded)
+      _.each(ads, _.bind(function(ad) {
+        if (supportsPoddedAds(ad.version) && typeof ad.sequence !== 'undefined' && _.isNumber(parseInt(ad.sequence))) {
+          //Assume sequences will start from 1
+          result.podded[+ad.sequence - 1] = ad;
```
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: What does the "+" prefix do in the context of the array access?

File: js/ad_manager_vast.js [hunk 4]
```diff
@@ -935,41 +993,113 @@ OO.Ads.manager(function(_, $) {
...
+            linear: { tracking: {}, ClickTracking: [] },
+            nonLinear: { tracking: {} } };
+          var params = {
+            adPodIndex : index + 1,
+            adPodLength : ads.length - nonLinearAdCount
+          };
+          this._mergeVastAdResult(ad, wrapperAds);
+          if (this._handleLinearAd(ad, adLoaded, params) || this._handleNonLinearAd(ad, adLoaded, params)) {
+            this.loaded = true;
+            this.trigger(this.READY, this);
```
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: Not sure, but if this is an ad pod should `` this.trigger(this.READY, this); `` be called after all the ads have been added to the time line and not after a single iteration?

File: js/ad_manager_vast.js [hunk 5]
```diff
@@ -697,24 +739,29 @@ OO.Ads.manager(function(_, $) {
...
       // filter our playable stream:
-      var firstNonLinearAd = _.find(this.inlineAd.ads, function(v){ return !_.isEmpty(v.nonLinear.url); }, this);
-      if (!firstNonLinearAd) { return false; }
-      var adURL = firstNonLinearAd.nonLinear.url;
-      this.vastAdUnit.streamUrl = adURL;
-      _.extend(this.vastAdUnit.data, firstNonLinearAd);
-      this.vastAdUnit.data.tracking = firstNonLinearAd.nonLinear.tracking;
-
-      if (this.vastAdUnit.streamUrl == null) {
+      if (_.isEmpty(ad.nonLinear.url)) { return false; }
```
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: Nitpicky, but for readability/debugging purposes I think t he return should be on a newline

File: js/ad_manager_vast.js [hunk 6]
```diff
@@ -697,24 +739,29 @@ OO.Ads.manager(function(_, $) {
...
+      params = params ? params : {};
+      var adURL = ad.nonLinear.url;
+      var vastAdUnit = { data: {}, vastUrl: this.vastUrl, maxBitrateStream: null };
```
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: Nitpicky, but I think it's slightly more readable to declare the object literal like:
```
var vastAdUnit = {
  data: {},
  vastUrl: this.vastUrl,
  maxBitrateStream: null
};
```

File: js/ad_manager_vast.js (hunk 1 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: The mergeVastAdResult function gets called inside handleAds a few lines down. ad.nonLinear is empty at this time.

File: js/ad_manager_vast.js (hunk 5 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: Agreed, will make this change.

File: js/ad_manager_vast.js (hunk 3 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: It turns it into a number. ad.sequence is originally a string at this point.

File: js/ad_manager_vast.js (hunk 4 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: Yep, sounds like it should be, but I don't believe it does anything here. Greg did want us to try to remove the dependency on the Emitter object since it doesn't work fully. I'll see if I could do that in this branch. If not will follow up on this.

File: js/ad_manager_vast.js [hunk 7]
```diff
@@ -979,63 +1109,25 @@ OO.Ads.manager(function(_, $) {
...
-        } else {
-          OO.log("Max wrapper depth reached.", this.currentDepth, OO.playerParams.maxVastWrapperDepth);
-          this.errorType = "tooManyWrapper";
-          this.trigger(this.ERROR, this);
-          failedAd();
+      } else {
+        var ad;
+        //TODO: Determine when we show standalone ads if podded ads are available
+        //Show podded ads if available
+        if(!_.isEmpty(vastAds.podded)) {
```
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: I feel this conditional along with the conditional on LINE 1129 might read off better like:
```
if(_.isEmpty(vastAds.podded)) {
	ad = vastAds.standalone[0];
	if (ad) {
		handleAds([ad], adLoaded);
	}
}
else {
	handleAds(vastAds.podded, adLoaded);
}
```
that way the "branch"/flow for the standalone logic is grouped together.

File: js/ad_manager_vast.js (hunk 1 above)
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: Ahh okay, oops

File: js/ad_manager_vast.js (hunk 3 above)
Comment (dchauOoyala) [inline js/ad_manager_vast.js]: Ahh, I see

File: js/ad_manager_vast.js (hunk 7 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: Will make this change.

File: js/ad_manager_template.js [hunk 8]
```diff
@@ -198,6 +199,15 @@ OO.Ads.manager(function(_, $) {
...
     };
 
+    /**
+     * <i>Optional.</i><br/>
+     * Called when the player detects start of ad video playback.
+     * @method AdManager#adVideoPlaying
+     * @public
+     */
+    this.adVideoPlaying = function() {
+
```
Comment (gfrankOoyala) [inline js/ad_manager_template.js]: Why do we need this if it is empty?

File: js/ad_manager_vast.js [hunk 9]
```diff
@@ -94,11 +87,20 @@ OO.Ads.manager(function(_, $) {
       'close' ];
 
     /**
-     * Helper function to verify that the xml has a valid vast ad in it and that it is a valid xml.
+     * Helper function to verify that XML is valid
      * @param {xml} vastXML Contains the vast ad data to be parsed.
      * @returns {boolean} Returns true if the xml is valid otherwise it returns false.
      */
-    var isValidVastXML = _.bind(function(vastXML) {
+    this.isValidVastXML = _.bind(function(vastXML) {
```
Comment (gfrankOoyala) [inline js/ad_manager_vast.js]: Public functions don't need _.bind

File: js/ad_manager_template.js (hunk 8 above)
Comment (aeng7) [inline js/ad_manager_template.js]: This is the ad manager template and will be used as a reference for people who want to implement their own ad managers.

File: js/ad_manager_vast.js [hunk 10]
```diff
@@ -406,22 +449,45 @@ OO.Ads.manager(function(_, $) {
...
+    };
+
+    /**
+     *
+     * @param ad
+     */
+    this.endAd = function(ad) {
       if (ad) {
         if (ad.isLinear) {
+          var endOfPod = ad.ad.adPodIndex === ad.ad.adPodLength;
```
Comment (gfrankOoyala) [inline js/ad_manager_vast.js]: Can me make this an if instead, makes it easier to read.

File: js/ad_manager_vast.js (hunk 9 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: Got it, will make this change.

File: js/ad_manager_vast.js [hunk 11]
```diff
@@ -442,48 +509,33 @@ OO.Ads.manager(function(_, $) {
...
       {
         type = this.amc.ADTYPE.NONLINEAR_OVERLAY;
-        duration = ad.data.nonLinear.Duration ?  OO.timeStringToSeconds(ad.data.nonLinear.Duration) : 0;
+        duration = metadata.data.nonLinear.Duration ?  OO.timeStringToSeconds(metadata.data.nonLinear.Duration) : 0;
       }
-      var positionSeconds = adLoaded.time/1000;
 
-      // Save the stream data for use by VideoController
-      var streams = {};
-      var linearAd = ad.data.linear;
```
Comment (gfrankOoyala) [inline js/ad_manager_vast.js]: Why was this removed? We need to build the streamData object ot have all the supported streams listed so the VTC can pick the right one.

File: js/ad_manager_vast.js [hunk 12]
```diff
@@ -636,11 +688,10 @@ OO.Ads.manager(function(_, $) {
     this._onVastError = function() {
       this.errorType = 'directAjaxFailed';
       this._ajax(this._getProxyUrl(), this._onFinalError, 'script');
-      this.trigger(this.ERROR, this);
```
Comment (gfrankOoyala) [inline js/ad_manager_vast.js]: Can we log an error here please

File: js/ad_manager_vast.js [hunk 13]
```diff
@@ -648,7 +699,6 @@ OO.Ads.manager(function(_, $) {
      */
     this._onFinalError = function() {
       this.errorType = "proxyAjaxFailed";
-      this.trigger(this.ERROR, this);
```
Comment (gfrankOoyala) [inline js/ad_manager_vast.js]: log here as well

File: js/ad_manager_vast.js [hunk 14]
```diff
@@ -935,41 +1018,171 @@ OO.Ads.manager(function(_, $) {
...
+        }, this));
+
+        handled = addToTimeline(adUnits[0]);
+      }
+
+      if (handled) {
+        this.loaded = true;
+      } else {
+        this.errorType = "noAd";
+        failedAd();
```
Comment (gfrankOoyala) [inline js/ad_manager_vast.js]: Can we log when this happens.

File: js/ad_manager_vast.js (hunk 11 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: This was moved to inside the _handleLinearAd function.

File: js/ad_manager_vast.js (hunk 12 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: This will be part of the error handling code that Douglas will be checking in. Should I include a temporary one for now?

File: js/ad_manager_vast.js (hunk 10 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: Will do.

**Comment by aeng7** [comment]
I've found a bug with replays, please do not merge yet.

File: js/ad_manager_template.js [hunk 15]
```diff
@@ -129,8 +129,9 @@ OO.Ads.manager(function(_, $) {
      * @method AdManager#cancelAd
      * @public
      * @param {object} ad The ad object to cancel
+     * @param {object} params An object containing information about the cancellation
```
Comment (angieisrani) [inline js/ad_manager_template.js]: What kind of information?  Is that documented anywhere?

File: js/ad_manager_vast.js [hunk 16]
```diff
@@ -12,7 +12,6 @@ require("../html5-common/js/utils/InitModules/InitOOPlayerParamsDefault.js");
 
 require("../html5-common/js/utils/constants.js");
 require("../html5-common/js/utils/utils.js");
-require("../html5-common/js/classes/emitter.js");
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: woo!

File: js/ad_manager_vast.js [hunk 17]
```diff
@@ -40,23 +37,18 @@ OO.Ads.manager(function(_, $) {
...
    * @property {object} lastOverlayAd Contains the ad information for the overlay that was displayed before it was removed.
    * This is used so we know what to add back to the screen after the video ad is done and the main video hasn't ended.
-   * @property {string} ERROR Constant used when triggering an error to indicate it was a Vast error
-   * @property {string} READY Constant used to trigger an event to indicate that the vast ad is ready for use
    * @property {string} VAST_AD_CONTAINER Constant used to keep track of the Vast Ad container div/layer that is used to
    * show the ads
    * @property {object} currentAdBeingLoaded Stores the ad data of the ad that is currently being loaded
-   * @property {object} wrapperAds Is used to keep track of the analytic and clickthrough info of an ad
    */
-  var Vast = OO.inherit(OO.Emitter, function() {
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: woo!

File: js/ad_manager_vast.js [hunk 18]
```diff
@@ -235,8 +264,32 @@ OO.Ads.manager(function(_, $) {
...
      */
     var failedAd = _.bind(function() {
-      // TODO: Do not destory the whole ad manager if one ad fails!
-      this.destroy();
+      var metadata = null;
+      var badAd = currentAd;
+      currentAd = null;
+
+      if (badAd) {
+        if(badAd.ad.fallbackAd) {
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: might need to check if badAd.ad?

File: js/ad_manager_vast.js [hunk 19]
```diff
@@ -294,6 +347,22 @@ OO.Ads.manager(function(_, $) {
     };
 
     /**
+     * Called when the ad starts playback.
+     * @public
+     * @method Vast#adVideoPlaying
+     */
+    this.adVideoPlaying = function() {
+      if (currentAd && currentAd.ad.nextAdInPod) {
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: if currentAd.ad?

File: js/ad_manager_vast.js [hunk 20]
```diff
@@ -294,6 +347,22 @@ OO.Ads.manager(function(_, $) {
...
+        var metadata = currentAd.ad.nextAdInPod;
+        var ad = generateAd(metadata);
+
+        if (metadata.streamUrl != null || (ad.adType == this.amc.ADTYPE.LINEAR_VIDEO && !_.isEmpty(metadata.streams))) {
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: if metadata?

Comment (angieisrani) [inline js/ad_manager_vast.js]: What is the difference between streamUrl and streams

File: js/ad_manager_vast.js [hunk 21]
```diff
@@ -235,8 +264,32 @@ OO.Ads.manager(function(_, $) {
...
+      if (badAd) {
+        if(badAd.ad.fallbackAd) {
+          metadata = badAd.ad.fallbackAd;
+        }
+        //notify amc of the end of the failed ad
+        if (badAd.isLinear) {
+          this.amc.notifyLinearAdEnded(badAd.id);
+        } else {
+          this.lastOverlayAd = null;
+          this.amc.notifyNonlinearAdEnded(badAd.id);
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: Can you call adCompletedCallback?

File: js/ad_manager_vast.js [hunk 22]
```diff
@@ -406,22 +486,43 @@ OO.Ads.manager(function(_, $) {
...
+        if (ad.isLinear) {
+          this.adVideoEnded();
+        } else {
+          this.endAd(ad);
+        }
+      }
+    };
+
+    /**
+     *
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: needs jsdoc
Is this public because AMF is calling it?  Does it need to be public?

Comment (angieisrani) [inline js/ad_manager_vast.js]: same for the functions that were made public

File: js/ad_manager_vast.js [hunk 23]
```diff
@@ -671,80 +770,101 @@ OO.Ads.manager(function(_, $) {
      *  If a linear ad is found, then it is parsed and sent to be added to the time via addToTimeLine.
      * @public
      * @method Vast#_handleLinearAd
+     * @param {object} ad The ad object
      * @param {object} adLoaded The ad that was loaded
-     * @returns {boolean} True if the ad was loaded and a stream was found; else false.
+     * @param {object} params Additional params
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: what kind of params

File: js/ad_manager_vast.js [hunk 24]
```diff
@@ -671,80 +770,101 @@ OO.Ads.manager(function(_, $) {
...
     };
 
     /**
      * If a non-linear Ad is found then it is parsed and added to the timeline via the addToTimeline function.
      * @public
      * @method Vast#_handleNonLinearAd
+     * @param {object} ad The ad object
      * @param {object} adLoaded The ad that was loaded
-     * @returns {boolean} True if the load was successful and a stream was found otherwise false.
+     * @param {object} params Additional params
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: what kinds

File: js/ad_manager_vast.js [hunk 25]
```diff
@@ -671,80 +770,101 @@ OO.Ads.manager(function(_, $) {
...
+     * @returns {object} The ad unit object ready to be added to the timeline
      */
-    this._handleNonLinearAd = function(adLoaded) {
+    this._handleNonLinearAd = function(ad, adLoaded, params) {
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: this function is public but is preceded by an underscore.  Is it meant to have been private?

File: js/freewheel.js [hunk 26]
```diff
@@ -460,9 +460,9 @@ OO.Ads.manager(function(_, $) {
             slotEndedCallbacks[ad.ad.getCustomId()] = _.bind(function(adId) {
                 amc.notifyPodEnded(adId);
               }, this, ad.id);
-            adStartedCallbacks[ad.ad.getCustomId()] = _.bind(function(details) {
-                amc.notifyLinearAdStarted(this.name, details);
-              }, this);
+            adStartedCallbacks[ad.ad.getCustomId()] = _.bind(function(adId, details) {
```
Comment (angieisrani) [inline js/freewheel.js]: is this new parameter already being used?  i don't see any other changes such as to the function calls

File: test/unit-test-helpers/mock_amc.js [hunk 27]
```diff
@@ -20,6 +20,12 @@ fake_amc = function() {
     LINEAR_VIDEO : "linearVideo",
     COMPANION : "companion"
   };
+  this.AD_CANCEL_CODE = {
+    SKIPPED : "skipped",
+    TIMEOUT : "timeout",
+    ERROR : "error",
+    STREAM_ENDED : "streamEnded"
+  };
```
Comment (angieisrani) [inline test/unit-test-helpers/mock_amc.js]: i am assuming this was already defined in real AMC?

Comment (aeng7) [inline test/unit-test-helpers/mock_amc.js]: Yep, it's defined in the pbi-741 branch, which has not been merged yet.

File: js/ad_manager_vast.js (hunk 20 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: streamUrl is used for nonlinear ads and streams is used for linear ads. This is the convention that was used prior to my changes.

File: js/ad_manager_vast.js (hunk 22 above)
Comment (aeng7) [inline js/ad_manager_vast.js]: Yep, the AMF calls cancelAd.

File: js/freewheel.js (hunk 26 above)
Comment (aeng7) [inline js/freewheel.js]: Before, we were calling notifyLinearAdStarted with the name of the ad manager rather than the ad id. This is to fix that issue.

Comment (angieisrani) [inline js/freewheel.js]: sounds good

File: test/unit-test-helpers/mock_amc.js (hunk 27 above)
Comment (angieisrani) [inline test/unit-test-helpers/mock_amc.js]: sounds good

File: js/ad_manager_vast.js (hunk 20 above)
Comment (angieisrani) [inline js/ad_manager_vast.js]: sounds good

File: js/ad_manager_template.js (hunk 15 above)
Comment (aeng7) [inline js/ad_manager_template.js]: Added documentation to the template and this plugin.

Comment (angieisrani) [inline js/ad_manager_template.js]: could i ask you to make the tiny change of adding something to the JSDoc like this:
"... about the cancellation.  It will include the following fields:"

Comment (aeng7) [inline js/ad_manager_template.js]: Sure, will do.

File: js/ad_manager_vast.js [hunk 28]
```diff
@@ -406,22 +488,48 @@ OO.Ads.manager(function(_, $) {
...
+
+    /**
+     * Ends an ad. Notifies the AMC about the end of the ad. If it is the last linear ad in the pod,
+     * will also notify the AMC of the end of the ad pod.
+     * @public
+     * @method Vast#endAd
+     * @param {object} ad The ad to end
+     * @param {boolean} failed If true, the ending of this ad was caused by a failure
+     */
+    this.endAd = function(ad, failed) {
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: Since AMF is not calling endAd, why is it public?

File: js/ad_manager_vast.js [hunk 29]
```diff
@@ -669,82 +775,107 @@ OO.Ads.manager(function(_, $) {
...
-     * @public
+     * @private
      * @method Vast#_handleLinearAd
+     * @param {object} ad The ad object
      * @param {object} adLoaded The ad that was loaded
-     * @returns {boolean} True if the ad was loaded and a stream was found; else false.
+     * @param {object} params Additional parameters associated with this ad
+     *                        adPodIndex : the index of the ad pod this ad is housed in
+     *                        adPodLength : the total number of ads in the ad pod this ad is housed in
+     * @returns {object} The ad unit object ready to be added to the timeline
```
Comment (angieisrani) [inline js/ad_manager_vast.js]: seems to return object or boolean

**Comment by angieisrani** [comment]
LGTM
